### PR TITLE
format: report total steps correctly in progress bar

### LINKED ISSUE #1408: Emit testCase messages before test run starts
Currently cucumber-js emits `testCase` messages right before each test case will be run. The CCK defines that they should be emitted together at the start of the test run.

(We're temporarily reordering the envelopes in our CCK test instrumentation at the moment.)

### LINKED ISSUE #1579: Steps counter show wrong values
Hello, since I upgrade to v7 I have this issue (that I didn't have in v5.1):

The counter on the progress bar when using in combinaison with tags show erroneous max value, and the progress-bar size is also not adapted.

**Step to reproduce:**
By your own:
Run cucumber-js with options: `--format progress-bar --tags <tag>`

**Minimal reproducible example:**
https://github.com/Florent-Bouisset/bug-reproduce-cucumber/
> npm run test

**Expected result:**
When the test is over the progress bar should show 5/5

**Actual result:** 
When the test is over the progress bar show 5/15 
15 is the number of step it should have done without the --tag option.

image

### PR DESCRIPTION
**Pull request by davidjgoss**
- Fixes #1408 
- Fixes #1579

Most of this is a refactor so that `testCase` messages are emitted upfront at the start of the run as opposed to right before each test case is run. This makes it possible for the progress bar formatter to base the total step count on test cases rather than pickles (which could be incorrect when e.g. using a tag filter).

`messages.TestCase` is now more or less the source of truth for test cases to be run - we have a bit less internal duplication and mapping of that data now e.g. `ITestStep` is gone.

`pickle_runner` is renamed to `test_case_runner` which seems more representative of what it does.

With the parallel runtime, previously the worker would emit `testCase` messages and then the coordinator would remap the ids for hooks and steps to be correct per the coordinator's support code library. With this change we flip it around, so the coordinator pushes down the "canonical" hook and step ids when initialising the worker, and the worker's support code library adopts those when finalising. This means messages can just be emitted without any remapping.

(Sorry for the big PR!)

### CODE REVIEW

**Review by aurelien-reeves — COMMENTED** [review]
Why did you keep assembleTestCases a function and not a class?

The whole PR looks nice, great job. After a first look it seems to be good 👍

File: features/core.feature [hunk 1]
```diff
@@ -1,3 +1,4 @@
+@dg
```
[the remark this answers is not in the document]
Comment (davidjgoss) [reply features/core.feature:1]: Good catch thanks!

File: CHANGELOG.md [hunk 2]
```diff
@@ -16,6 +16,8 @@ Please see [CONTRIBUTING.md](https://github.com/cucumber/cucumber/blob/master/CO
 
 ### Changed
 
+* All `testCase` messages now emitted upfront at the start of the run ([#1408](https://github.com/cucumber/cucumber-js/issues/1408)
+  [#1669](https://github.com/cucumber/cucumber-js/pull/1669))
```
[the remark this answers is not in the document]
Comment (davidjgoss) [reply CHANGELOG.md]: Good point, added a note that it's about formatters

**Comment by davidjgoss** [comment]
> Why did you export assembleTestCases a function and not a class?

it just does one thing and doesn't need to hold any state, so a class would seem unnecessary.

**Comment by aurelien-reeves** [comment]
> 
> 
> > Why did you export assembleTestCases a function and not a class?
> 
> It just does one thing and doesn't need to hold any state, so a class would seem unnecessary.

Ok, thanks for clarifying it :)

**Review by aslakhellesoy — APPROVED** [review]
Excellent stuff @davidjgoss !

**Comment by trenowden** [comment]
Great to see this fixed! Any idea when this will be published to npm?

**Comment by aurelien-reeves** [comment]
I am pretty sure this will be released "soon". But we cannot know for sure actually.

File: src/runtime/parallel/coordinator.ts [hunk 3]
```diff
@@ -162,6 +121,17 @@ export default class Coordinator {
         filterStacktraces: this.options.filterStacktraces,
         supportCodePaths: this.supportCodePaths,
         supportCodeRequiredModules: this.supportCodeRequiredModules,
+        supportCodeIds: {
```
Comment (charlierudolph) [inline src/runtime/parallel/coordinator.ts:124]: Cool. Nice refactor

**Comment by charlierudolph** [comment]
Thanks for doing this @davidjgoss!

**Pull request merged**